src/combinations.rs

```diff
@@ -120,9 +120,51 @@ impl<I> Iterator for Combinations<I>
         // Create result vector based on the indices
         Some(self.indices.iter().map(|i| self.pool[*i].clone()).collect())
     }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let (mut low, mut upp) = self.pool.size_hint();
+        low = remaining_for(low, self.first, &self.indices).unwrap_or(usize::MAX);
+        upp = upp.and_then(|upp| remaining_for(upp, self.first, &self.indices));
+        (low, upp)
+    }
+
+    fn count(self) -> usize {
+        let Self { indices, pool, first } = self;
+        let n = pool.len() + pool.it.count();
+        remaining_for(n, first, &indices).expect("Iterator count greater than usize::MAX")
+    }
 }
 
 impl<I> FusedIterator for Combinations<I>
     where I: Iterator,
           I::Item: Clone
 {}
+
+pub(crate) fn checked_binomial(mut n: usize, mut k: usize) -> Option<usize> {
+    if n < k {
+        return Some(0);
+    }
+    // `factorial(n) / factorial(n - k) / factorial(k)` but trying to avoid it overflows:
+    k = (n - k).min(k); // symmetry
+    let mut c = 1;
+    for i in 1..=k {
+        c = (c / i).checked_mul(n)?.checked_add((c % i).checked_mul(n)? / i)?;
+        n -= 1;
+    }
+    Some(c)
+}
+
+/// For a given size `n`, return the count of remaining combinations or None if it would overflow.
+fn remaining_for(n: usize, first: bool, indices: &[usize]) -> Option<usize> {
+    let k = indices.len();
+    if first {
+        checked_binomial(n, k)
+    } else {
+        indices
+            .iter()
+            .enumerate()
+            .fold(Some(0), |sum, (k0, n0)| {
+                sum.and_then(|s| s.checked_add(checked_binomial(n - 1 - *n0, k - k0)?))
+            })
```

Comment:
Thanks for your explanations.

Please add these links in the comments for the next reader:
* https://en.wikipedia.org/wiki/Combinatorial_number_system
* http://www.site.uottawa.ca/~lucia/courses/5165-09/GenCombObj.pdf

And maybe the following explanation within "The combinations generated after the current one can be counted by counting as follows:
* The subsequent combinations that differ in indices[0]: If subsequent combinations differ in indices[0], then their value for indices[0] must be at least 1 greater than the current indices[0]. As indices is strictly monotonically sorted, this means we can effectively choose k values from (n - 1 - indices[0]), leading to binomial(n-1-indices[0], k) possibilities.
* The subsequent combinations with same indices[0], but differing indices[1]: Here we can choose k-1 values from (n-1-indices[1]) values, leading to binomial(n-1-indices[1], k-1) possibilities.
* (...)
* The subsequent combinations with same indices[0..=i], but differing indices[i]: Here we can choose k-i values from (n-1-indices[i]) values: binomial(n-1-indices[i], k-i).
Since subsequent combinations can in any index, we must sum up the aforementioned binomial coefficients."

Please rename `k0` to `i` so that it matches the comment, and comment that `n0` resembles `indices[i]`.

Reply:
Oh my! I missed that wikipedia link. Done!